makerdao-bridge-invariant/src/agent.ts

```diff
@@ -0,0 +1,137 @@
+import { Finding, TransactionEvent, ethers, getEthersProvider, AlertsResponse } from 'forta-agent';
+import L1Escrow from './L1Escrow.agent';
+
+import {
+  l1ArbitrumGateway,
+  l1ArbitrumEscrowAddress,
+  l1ArbitrumEvent,
+  l1DAIAddress,
+  l1DAI_Decimals,
+  l1OptimismEvent,
+  l1OptimismEscrowAddress,
+  l1OptimismGateway,
+  transferEvent,
+  totalSupplyABI,
+} from './config.abi';
+import { arbitrumL1EscrowFinding, optimismL1EscrowFinding } from './l1Findings';
+import { NetworkManager, ProviderCache } from 'forta-agent-tools';
+import { createAddress } from 'forta-agent-tools/lib/';
+import { getNormalizedAmount, getTokenSupply } from './utils';
+import { getL1Alerts, networkData } from './L2helper';
+
+export function provideHandleTransaction(
+  provider: ethers.providers.Provider,
+  getL1Alerts: (alertId: string, blockTimestamp: number) => Promise<AlertsResponse>
+) {
+  return async function handleTransaction(txEvent: TransactionEvent) {
+    let findings: Finding[] = [];
+
+    const { chainId } = await provider.getNetwork();
+
+    if (chainId === 1) {
+      const arbitrumL1Escrow = new L1Escrow(
+        'Arbitrum',
+        l1ArbitrumEvent,
+        l1ArbitrumEscrowAddress,
+        l1ArbitrumGateway,
+        l1DAIAddress,
+        l1DAI_Decimals,
+        provider,
+        arbitrumL1EscrowFinding
+      );
+
+      const optimismL1Escrow = new L1Escrow(
+        'Optimism',
+        l1OptimismEvent,
+        l1OptimismEscrowAddress,
+        l1OptimismGateway,
+        l1DAIAddress,
+        l1DAI_Decimals,
+        provider,
+        optimismL1EscrowFinding
+      );
+
+      findings = (
+        await Promise.all([
+          await arbitrumL1Escrow.handleTransaction(txEvent),
+          await optimismL1Escrow.handleTransaction(txEvent),
+        ])
+      ).flat();
+    } else {
+      const networkManager = new NetworkManager(networkData);
+
+      await networkManager.init(provider);
```

Comment:
This is an expensive call to be made in every transaction. You can add a `provideInitialize` block which is called only once.